setup.py

```diff
@@ -1,8 +1,34 @@
 # Need setuptools even though it isn't used - loads some plugins.
 import os
-from distutils.core import setup
+import sys
 
-from setuptools import find_packages  # noqa: F401
+from setuptools import find_packages, setup  # noqa: F401
+
+
+# Taken from Numba
+def _guard_python_version(max_python):
+    version_module = None
+    try:
+        from packaging import version as version_module
+    except ImportError:
+        try:
+            from setuptools._vendor.packaging import version as version_module
+        except ImportError:
+            pass
+
+    if version_module is None:
+        return
+
+    current_python = version_module.parse(".".join(map(str, sys.version_info[:3])))
+    max_python = version_module.parse(max_python)
+
+    if not current_python < max_python:
+        raise RuntimeError(
+            f"Cannot install on Python version {current_python} as Python {max_python}+ is not yet supported."
+        )
+
+
+_guard_python_version(max_python="3.13")
```

Comment:
There may be AST related things in `func_adl` that make it (reasonably!) fragile to CPython version changes, so to still guard against this without using a `requires-python` version cap that eventually causes issues, use a version guard like Numba does.

c.f.https://github.com/iris-hep/qastle/pull/72#discussion_r1529696709 for additional context.